integrations/ray.md

```diff
@@ -0,0 +1,350 @@
+```
+
+Key takeways from the example:
+
+- import `ray` module
+- import `RayPipeline` (from `ray_haystack`) instead of `Pipeline` class from `haystack`
+- before running the pipeline [local ray cluster](https://docs.ray.io/en/latest/ray-core/starting-ray.html#start-ray-init) is started with explicit `ray.init()` call
+
+Under the hood `RyaPipeline` creates actors for each component in the pipeline and runs it in distributed manner until no components left to run. By default `RyaPipeline` blocks until pipeline finishes its execution.
+
+### Read pipeline events
+
+In some cases you would want to asynchronously react to particular pipeline execution points:
+
+- when pipeline starts
+- before component runs
+- after component finishes
+- after pipeline finishes
+
+Internally `RayPipeline` creates an instance of [Ray Queue](https://docs.ray.io/en/latest/ray-core/api/doc/ray.util.queue.Queue.html) where such events are being stored and can be consumed from.
+
+Except standard `run` method `RayPipeline` provides a method called `run_nowait` which returns pipeline execution result without blocking current logic. We can use `run_nowait` to iterate over pipeline events, e.g.
+
+```python
+result = pipeline.run_nowait(pipeline_inputs)
+
+for pipeline_event in result.pipeline_events_sync():
+    print(
+        f"\n>>> [{pipeline_event.time}] Source: {pipeline_event.source} | Type: {pipeline_event.type} | Data={pipeline_event.data}"
+    )
+```
+
+### Component Serialization
+
+If you run native Haystack pipeline locally component remain in the same python process and there is no reason to care about distributed setup. When RayPipeline runs in distributed manner it should be able to [serialize](https://docs.ray.io/en/latest/ray-core/objects/serialization.html) components before they end up in remote task or actor.
+
+`ray-haystack` package relies on Haystack's standard `to_dict` and `from_dict` methods to serialize and deserialize components respectively.
+Please refer to main [documentation](https://github.com/prosto/ray-haystack?tab=readme-ov-file#component-serialization) to see a detailed explanation and features which handle edge cases.
+
+### DocumentStore with Ray
+
+When you use [InMemoryDocumentStore](https://docs.haystack.deepset.ai/docs/inmemorydocumentstore) or any DocumentStore which runs in-memory with `RayPipeline` you will stumble upon an apparent issue: in distributed environment such DocumentStore will fail to operate as components which reference the store will not point to single instance but rather a copy of it.
+
+`ray-haystack` package provides a wrapper around `InMemoryDocumentStore` by implementing a proxy pattern so that only a single instance of `InMemoryDocumentStore` across Ray cluster is present. With that pipeline components could share a single store. Use `RayInMemoryDocumentStore`, `RayInMemoryEmbeddingRetriever` or `RayInMemoryBM25Retriever` in case you need in-memory document store in your Ray pipelines.
```

Comment:
```suggestion
When you use [InMemoryDocumentStore](https://docs.haystack.deepset.ai/docs/inmemorydocumentstore) or any DocumentStore that runs in memory with `RayPipeline`, you will stumble upon an apparent issue: in a distributed environment, these document stores running in memory will fail to operate as components that reference the store will not point to a single instance but rather a copy of it.

`ray-haystack` package provides a wrapper around `InMemoryDocumentStore` by implementing a proxy pattern so that only a single instance of `InMemoryDocumentStore` across Ray cluster is present. With that, components can share a single store. Use `RayInMemoryDocumentStore`, `RayInMemoryEmbeddingRetriever` or `RayInMemoryBM25Retriever` in case you need in-memory document store in your Ray pipelines.
```